core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala

```diff
@@ -0,0 +1,128 @@
+    refreshInterval: Option[FiniteDuration],
+    fetchSize: Int,
+    preparedStatement: PreparedStatement,
+    session: Session)
+    extends GraphStage[SourceShape[String]] {
+
+  import AllPersistenceIdsStage._
+
+  val out: Outlet[String] = Outlet("AllPersistenceIds.out")
+
+  val shape: SourceShape[String] = SourceShape(out)
+
+  def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogic(shape) with StageLogging with OutHandler {
+
+      implicit def executionContext: ExecutionContextExecutor = materializer.executionContext
+
+      private var queryInProgress = false
+      private var knownPersistenceIds: Set[String] = Set.empty
+      private var maybeResultSet = Option.empty[ResultSet]
+      private var buffer: Queue[String] = Queue.empty[String]
+
+      private val queryCallback: AsyncCallback[ResultSet] =
+        getAsyncCallback[ResultSet] { rs =>
+          queryInProgress = false
+          maybeResultSet = Some(rs)
+          val available = rs.getAvailableWithoutFetching
+          for (_ <- 1 to available) {
+            val s = rs.one().getString("persistence_id")
+            if (!knownPersistenceIds.contains(s)) {
+              buffer = buffer.enqueue(s)
+              knownPersistenceIds += s
+            }
+          }
+          flush()
+          if (refreshInterval.isEmpty && buffer.isEmpty && rs.isExhausted) {
+            complete(out)
+          }
+        }
+
+      private def query(): Unit =
+        if (!queryInProgress) {
+          queryInProgress = true
+          val boundStatement = preparedStatement.bind()
+          boundStatement.setFetchSize(fetchSize)
+          session.executeAsync(boundStatement).asScala.foreach(queryCallback.invoke)
+        }
+
+      private def flush(): Unit = {
+        while (buffer.nonEmpty && isAvailable(out)) {
+          val (s, newBuffer) = buffer.dequeue
+          buffer = newBuffer
+          push(out, s)
+        }
+      }
+
+      override def preStart(): Unit = {
+        query()
+        refreshInterval.foreach { interval =>
+          val initial =
+            if (interval >= 2.seconds)
+              (interval / 2) + ThreadLocalRandom.current().nextLong(interval.toMillis / 2).millis
+            else interval
+          schedulePeriodicallyWithInitialDelay(Continue, initial, interval)
+        }
+      }
+
+      override def onTimer(timerKey: Any): Unit = {
+        timerKey match {
+          case Continue =>
+            query()
+
+          case _ =>
+        }
+      }
+
+      def onPull(): Unit = {
+        flush()
+        if (buffer.isEmpty && isAvailable(out)) {
+          maybeResultSet match {
+            case None =>
+              query()
+
+            case Some(rs) =>
+              if (refreshInterval.isEmpty && rs.isExhausted) {
+                complete(out)
+              } else {
+                rs.fetchMoreResults().asScala.foreach(queryCallback.invoke)
```

Comment:
Should this also be guarded by `queryInProgress` to avoid calling `fetchMoreResults` in parallel?

Reply:
Yep I think so, I'll add that change then merge this!